pkg/network/tracer/ebpf_conntracker.go

```diff
@@ -0,0 +1,294 @@
+	var err error
+	e.consumer, err = netlink.NewConsumer(cfg.ProcRoot, cfg.ConntrackRateLimit, true)
+	if err != nil {
+		return err
+	}
+	defer e.consumer.Stop()
+
+	if err := e.loadInitialState(ctx, e.consumer.DumpTable(unix.AF_INET)); err != nil {
+		return err
+	}
+	if err := e.loadInitialState(ctx, e.consumer.DumpTable(unix.AF_INET6)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *ebpfConntracker) loadInitialState(ctx context.Context, events <-chan netlink.Event) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case ev, ok := <-events:
+			if !ok {
+				return nil
+			}
+			e.processEvent(ev)
+		}
+	}
+}
+
+func (e *ebpfConntracker) processEvent(ev netlink.Event) {
+	conns := netlink.DecodeAndReleaseEvent(ev)
+	for _, c := range conns {
+		if netlink.IsNAT(c) {
+			log.Tracef("initial conntrack %s", c)
+			src := formatKey(uint32(c.NetNS), c.Origin)
+			dst := formatKey(uint32(c.NetNS), c.Reply)
+			if src != nil && dst != nil {
+				if err := e.addTranslation(src, dst); err != nil {
+					log.Warnf("error adding initial conntrack entry to ebpf map: %s", err)
+				}
+				if err := e.addTranslation(dst, src); err != nil {
+					log.Warnf("error adding initial conntrack entry to ebpf map: %s", err)
+				}
+			}
+		}
+	}
+}
+
+func (e *ebpfConntracker) addTranslation(src *ConnTuple, dst *ConnTuple) error {
+	if err := e.ctMap.Update(unsafe.Pointer(src), unsafe.Pointer(dst), ebpf.UpdateNoExist); err != nil && !errors.Is(err, ebpf.ErrKeyExist) {
+		return err
+	}
+	return nil
+}
+
+func formatKey(netns uint32, tuple *ct.IPTuple) *ConnTuple {
+	var proto network.ConnectionType
+	switch *tuple.Proto.Number {
+	case unix.IPPROTO_TCP:
+		proto = network.TCP
+	case unix.IPPROTO_UDP:
+		proto = network.UDP
+	default:
+		return nil
+	}
+
+	return newConnTuple(0,
+		netns,
+		util.AddressFromNetIP(*tuple.Src),
+		util.AddressFromNetIP(*tuple.Dst),
+		*tuple.Proto.SrcPort,
+		*tuple.Proto.DstPort,
+		proto)
+}
+
+func (e *ebpfConntracker) GetTranslationForConn(stats network.ConnectionStats) *network.IPTranslation {
+	start := time.Now()
+	src := connTupleFromConnectionStats(&stats)
+	src.pid = 0
+	log.Tracef("looking up in conntrack: %s", src)
+
+	var dst ConnTuple
+	if err := e.ctMap.Lookup(unsafe.Pointer(src), unsafe.Pointer(&dst)); err != nil {
+		if !errors.Is(err, ebpf.ErrKeyNotExist) {
+			log.Warnf("error looking up connection in ebpf conntrack map: %s", err)
+		}
+		return nil
+	}
+
+	atomic.AddInt64(&e.stats.gets, 1)
+	atomic.AddInt64(&e.stats.getTotalTime, time.Now().Sub(start).Nanoseconds())
+	return &network.IPTranslation{
+		ReplSrcIP:   dst.SourceAddress(),
+		ReplDstIP:   dst.DestAddress(),
+		ReplSrcPort: uint16(dst.sport),
+		ReplDstPort: uint16(dst.dport),
+	}
+}
+
+func (e *ebpfConntracker) DeleteTranslation(stats network.ConnectionStats) {
+	start := time.Now()
+	key := connTupleFromConnectionStats(&stats)
+	key.pid = 0
+	if err := e.ctMap.Delete(unsafe.Pointer(key)); err != nil {
```

Comment:
Don't we also want to delete the translation of the translation? Since we add it in `addTranslation` above.

Reply:
Yes. Fixed.